### PR TITLE
Collapse some Source table columns into single JSONB column

### LINKED ISSUE #128: Add metadata storage column to source

### PR DESCRIPTION
**Pull request by acrellin**
Closes #128 & #154

### CODE REVIEW

File: skyportal/models.py [hunk 1]
```diff
@@ -112,24 +116,7 @@ class Source(Base):
 
     score = sa.Column(sa.Float, nullable=True)
 
-    ## pan-starrs
-    sgmag1 = sa.Column(sa.Float, nullable=True)
```
Comment (acrellin) [inline skyportal/models.py]: @profjsb these are the columns we're moving to the single `other_metadata` JSONB column. Any thoughts/suggestions?

File: skyportal/models.py [hunk 2]
```diff
@@ -112,24 +116,7 @@ class Source(Base):
...
-    srmag1 = sa.Column(sa.Float, nullable=True)
-    simag1 = sa.Column(sa.Float, nullable=True)
-    objectidps1 = sa.Column(sa.BigInteger, nullable=True)
-    sgscore1 = sa.Column(sa.Float, nullable=True)
-    distpsnr1 = sa.Column(sa.Float, nullable=True)
-
-    origin = sa.Column(sa.String, nullable=True)
-    modified = sa.Column(sa.DateTime, nullable=False,
```
Comment (profjsb) [inline skyportal/models.py]: should keep `modified` as it's a generic record for when the row is changed.

File: skyportal/models.py [hunk 3]
```diff
@@ -112,24 +116,7 @@ class Source(Base):
...
 
-    ## pan-starrs
-    sgmag1 = sa.Column(sa.Float, nullable=True)
-    srmag1 = sa.Column(sa.Float, nullable=True)
-    simag1 = sa.Column(sa.Float, nullable=True)
-    objectidps1 = sa.Column(sa.BigInteger, nullable=True)
-    sgscore1 = sa.Column(sa.Float, nullable=True)
-    distpsnr1 = sa.Column(sa.Float, nullable=True)
-
-    origin = sa.Column(sa.String, nullable=True)
```
Comment (profjsb) [inline skyportal/models.py]: `origin` is a good column to keep for provenance -- in fact, I'd suggest making this not nullable.

Comment (acrellin) [reply skyportal/models.py]: @profjsb these lines were actually just moved up a little bit, not removed

**Comment by acrellin** [comment]
@profjsb the columns that have been collapsed into the new matadata JSONB column are: 
```    sgmag1 = sa.Column(sa.Float, nullable=True)
    srmag1 = sa.Column(sa.Float, nullable=True)
    simag1 = sa.Column(sa.Float, nullable=True)
    objectidps1 = sa.Column(sa.BigInteger, nullable=True)
    sgscore1 = sa.Column(sa.Float, nullable=True)
    distpsnr1 = sa.Column(sa.Float, nullable=True)
    simbad_class = sa.Column(sa.Unicode, nullable=True, )
    simbad_info = sa.Column(JSONB, nullable=True)
    gaia_info = sa.Column(JSONB, nullable=True)
    tns_info = sa.Column(JSONB, nullable=True)
    tns_name = sa.Column(sa.Unicode, nullable=True)
```
Any feedback?

File: skyportal/models.py (hunk 3 above)
Comment (profjsb) [reply skyportal/models.py]: ok

**Comment by profjsb** [comment]
It's to move those ZTF-specific data to `other_metadata` JSONB field. But we should move the remainder of the data to `external_metadata` JSONB column. This can hold externally available metadata about that source that is not found in the survey from which that source originated. It can be structured like:

```json
external_metadata = 
 {"simbad": {"class": "RRL", "info": {...}},
  "TNS": {"name": "SN1987a", "info": {...}},
  "Gaia": {"info": {...}}
 }
``` 
So the lookups in the ZTF version of skyportal would be to `external_metadata.Gaia.info.Id` etc.

**Comment by acrellin** [comment]
> It's to move those ZTF-specific data to `other_metadata` JSONB field. But we should move the remainder of the data to `external_metadata` JSONB column. This can hold externally available metadata about that source that is not found in the survey from which that source originated. It can be structured like:
> 
> ```json
> external_metadata = 
>  {"simbad": {"class": "RRL", "info": {...}},
>   "TNS": {"name": "SN1987a", "info": {...}},
>   "Gaia": {"info": {...}}
>  }
> ```
> 
> So the lookups in the ZTF version of skyportal would be to `external_metadata.Gaia.info.Id` etc.

Thanks for the feedback - I'll change it accordingly.

**Comment by stefanv** [comment]
For `external_metadata`, those fields should be defined in the configuration file so that we can check such entries as they are uploaded.

Easiest way for now may be to use the same description as [OpenAPI](https://swagger.io/docs/specification/describing-parameters/).

**Comment by profjsb** [comment]
I'm definitely in favor of having application-specific/source-specific configs here.

File: skyportal/handlers/source.py [hunk 4]
```diff
@@ -206,10 +206,10 @@ def post(self):
             end_date = arrow.get(data['endDate'].strip())
             q = q.filter(Source.last_detected <= end_date)
         if data['simbadClass']:
-            q = q.filter(func.lower(Source.simbad_class) ==
-                         data['simbadClass'].lower())
+            q = q.filter(func.lower(Source.external_metadata['simbad']['class'].astext)
```
Comment (stefanv) [inline skyportal/handlers/source.py]: Does this work if `simbad` is not in `external_metadata`?

Comment (acrellin) [reply skyportal/handlers/source.py]: @stefanv yes, this works even when `simbad` is not in `external_metadata` for some or all sources.

**Comment by acrellin** [comment]
@profjsb do we still want to collapse ZTF-, Gaia-, Simbad- and TNS-specific fields into a JSONB column?

**Comment by dannygoldstein** [comment]
@acrellin yes, reviving this conversation.

**Comment by acrellin** [comment]
Closing in favor of https://github.com/skyportal/skyportal/pull/446

**Pull request closed without merging**